File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    answer = 0
    s = list(s)
    for i in range(len(s)):
        stack = []
        for j in range(len(s)):
            if stack:
                if stack[-1] == '(' and s[j] == ')':
                    stack.pop()
                elif stack[-1] == '[' and s[j] == ']':
                    stack.pop()
                elif stack[-1] == '{' and s[j] == '}':
                    stack.pop()
                else:
                    stack.append(s[j])
            else:
                stack.append(s[j])
        if not stack:
            answer += 1
        s.append(s.pop(0))
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>
#include <unordered_map>

using namespace std;

bool check(string target) {
    set<char> open = {'(','{','['};
    unordered_map<char,char> p = {
        {')','('},
        {'}','{'},
        {']','['}
    };
    
    vector<char> stack;
    for(char& t: target) {
        if (open.count(t)) { // 여는괄호
            stack.push_back(t);
        } else { // 닫는괄호
            if (!stack.empty() && stack.back() == p[t]) { // 쌍이 일치함
                stack.pop_back();
            } else {
                return false;
            }
        }
    }
    if (!stack.empty()) return false;
    return true;
}

int solution(string s) {
    int answer = 0;
    
    for(int x=0;x<s.length();x++) {
        string tmp = string(s.begin()+x, s.end()) + string(s.begin(), s.begin()+x);
        if(check(tmp)) answer++;
    }
    
    return answer;
}